logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageEnvelopeCreator.kt

```diff
@@ -0,0 +1,62 @@
+package com.wire.kalium.logic.feature.message
+
+import com.wire.kalium.cryptography.CryptoClientId
+import com.wire.kalium.cryptography.CryptoSessionId
+import com.wire.kalium.cryptography.ProteusClient
+import com.wire.kalium.cryptography.exceptions.ProteusException
+import com.wire.kalium.logic.CoreFailure
+import com.wire.kalium.logic.data.conversation.Recipient
+import com.wire.kalium.logic.data.message.ClientPayload
+import com.wire.kalium.logic.data.message.EncryptedMessageBlob
+import com.wire.kalium.logic.data.message.Message
+import com.wire.kalium.logic.data.message.MessageEnvelope
+import com.wire.kalium.logic.data.message.ProtoContent
+import com.wire.kalium.logic.data.message.ProtoContentMapper
+import com.wire.kalium.logic.data.message.RecipientEntry
+import com.wire.kalium.logic.functional.Either
+import com.wire.kalium.logic.functional.suspending
+import com.wire.kalium.cryptography.UserId as CryptoUserId
+
+interface MessageEnvelopeCreator {
+
+    suspend fun createOutgoingEnvelope(
+        recipients: List<Recipient>,
+        message: Message
+    ): Either<CoreFailure, MessageEnvelope>
+
+}
+
+class MessageEnvelopeCreatorImpl(
+    private val proteusClient: ProteusClient,
+    private val protoContentMapper: ProtoContentMapper
+) : MessageEnvelopeCreator {
+
+    override suspend fun createOutgoingEnvelope(
+        recipients: List<Recipient>,
+        message: Message
+    ): Either<CoreFailure, MessageEnvelope> = suspending {
+        val senderClientId = message.senderClientId
+        val content = protoContentMapper.encodeToProtobuf(ProtoContent(message.id, message.content))
+
+        recipients.foldToEitherWhileRight(mutableListOf<RecipientEntry>()) { recipient, recipientAccumulator ->
```

Comment:
should user self clients be added to the recipients' list?

Reply:
First: Go to sleep! It's 1:38 in the morning 💤 

Second: Yes. It's not the `MessageEnvelopeCreator`'s responsibility to judge that. But there will be an entity that will get all recipients (including other self user clients) and call the envelope creator with that.